servant-client/src/Servant/Client/Generic.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+module Servant.Client.Generic
+  ( ClientLike(..)
+  , genericMkClient
+  ) where
+
+import Generics.SOP   (Generic, I(..), NP(..), NS(Z), Rep, SOP(..), to)
+import Servant.API    ((:<|>)(..))
+import Servant.Client (ClientM)
+
+-- | This class allows us to match client structure with client functions
+-- produced with 'client' without explicit pattern-matching.
+--
+-- The client structure needs a 'Generics.SOP.Generic' instance.
+--
+-- Example:
+--
+-- type API
+--     = "foo" :> Capture "x" Int :> Get '[JSON] Int
+--  :<|> "bar" :> QueryParam "a" Char :> QueryParam "b" String :> Post '[JSON] [Int]
+--  :<|> Captre "nested" Int :> NestedAPI
+--
+-- type NestedAPI
+--     = Get '[JSON] String
+--  :<|> "baz" :> QueryParam "c" Char :> Post '[JSON] ()
+--
+-- data APIClient = APIClient
+--   { getFoo         :: Int -> ClientM Int
+--   , postBar        :: Maybe Char -> Maybe String -> ClientM [Int]
+--   , mkNestedClient :: Int -> NestedClient
+--   } deriving GHC.Generic
+--
+-- instance Generic.SOP.Generic APIClient
+-- instance (Client API ~ client) => ClientLike client APIClient
+--
+-- data NestedClient = NestedClient
+--  { getString ::  ClientM String
+--  , postBaz   :: Maybe Char -> ClientM ()
+--  } deriving GHC.Generic
+--
+-- instance Generic.SOP.Generic
+-- instance (Client NestedAPI ~ client) => ClientLike client NestedAPI
+--
+-- mkAPIClient :: APIClient
+-- mkAPIClient = mkClient (client (Proxy :: Proxy API))
+class ClientLike client custom where
+  mkClient :: client -> custom
```

Comment:
It make be that I'm missing something, but if we had:

```haskell
class (HasClient api) => ClientLike api custom where
   mkClient :: Client client -> custom
```

Then wouldn't we be able to say:

```haskell
instance ClientLike NestedAPI NestedClient
```
Instead of:
```haskell
instance (Client NestedAPI ~ client) => ClientLike client NestedClient
```

And maybe even get away with:

```haskell
data NestedClient = ... deriving (ClientLike NestedAPI)
```